### PR TITLE
feat(cse): add new microservice resource support

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
Support a new resource of CSE microservice:
- Support both secure authentication and insecure authentication request methods.

The dedicated microservice needs to use the service registry address of the microservice engine during the registration process, so I support a new Client function to request CSE services based on this address.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add new microservice resource support.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
PENDING
```

### CODE REVIEW

File: docs/resources/cse_microservice.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  Changing this will create a new microservice.
+
+* `level` - (Optional, String, ForceNew) Specifies the microservice level.
+  The valid values are **FRONT**, **MIDDLE**, and **BACK**. Changing this will create a new microservice.
+
+* `description` - (Optional, String, ForceNew) Specifies the description of the dedicated microservice.
+  The description can contian a maximum of `256` characters.
+  Changing this will create a new microservice.
+
+* `status` - (Optional, String, ForceNew) Specifies the microservice status. The valid values are **UP** and **DOWN**.
```
Comment (hmbig2) [inline docs/resources/cse_microservice.md]: forceNew will delete the instance, if cannot update the status,  change to computed attr is better.

File: docs/resources/cse_microservice.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  Changing this will create a new microservice.
+
+* `admin_user` - (Optional, String, ForceNew) Specifies the account name. The initial account name is **root**.
+  Required if the `auth_type` of engine is **RBAC**. Changing this will create a new microservice.
```
Comment (hmbig2) [inline docs/resources/cse_microservice.md:80]: no `auth_type` attribute

File: docs/resources/cse_microservice.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+* `admin_pass` - (Optional, String, ForceNew) Specifies the account password.
+  Required if the `auth_type` of engine is **RBAC**. Changing this will create a new microservice.
```
Comment (hmbig2) [inline docs/resources/cse_microservice.md:83]: no auth_type attribute

File: huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+			Status:      d.Get("status").(string),
+		},
+	}
+}
+
+func resourceMicroserviceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var token string
+	if user, ok := d.GetOk("admin_user"); ok {
+		var err error
+		client := NewCustomClient(d.Get("connect_address").(string), "v4")
```
Comment (hmbig2) [inline huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go]: why the version is 'v4' , not get from `version`

File: huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+	)
+
+	return diag.FromErr(mErr.ErrorOrNil())
+}
+
+func resourceMicroserviceDelete(_ context.Context, d *schema.ResourceData, _ interface{}) diag.Diagnostics {
+	var token string
+	if user, ok := d.GetOk("admin_user"); ok {
+		var err error
+		client := NewCustomClient(d.Get("connect_address").(string), "v4")
```
Comment (hmbig2) [inline huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go:173]: ditto

File: docs/resources/cse_microservice_instance.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+* `health_check` - (Optional, List, ForceNew) Specifies the health check configuration.
+  The [object](#microservice_instance_health_check) structure is documented below.
+  Changing this will create a new microservice instance.
+
+* `data_center` - (Optional, List, ForceNew) Specifies the data center configuration.
+  The [object](#microservice_instance_data_center) structure is documented below.
+  Changing this will create a new microservice instance.
+
+* `admin_user` - (Optional, String, ForceNew) Specifies the account name. The initial account name is **root**.
+  Required if the `auth_type` of engine is **RBAC**. Changing this will create a new microservice instance.
```
Comment (hmbig2) [inline docs/resources/cse_microservice_instance.md:115]: ditto

File: docs/resources/cse_microservice_instance.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+* `status` - The microservice instance status. The values supports **UP**, **DOWN**, **STARTING** and **OUTOFSERVICE**.
+
+## Import
+
+Microservices can be imported using related `connect_address`, `microservice_id` and their `id`, separated by a
+slash (/), e.g.
+
+```
+$ terraform import huaweicloud_cse_microservice.test https://124.70.26.32:30100/f14960ba495e03f59f85aacaaafbdef3fbff3f0d
```
Comment (hmbig2) [inline docs/resources/cse_microservice_instance.md]: s/huaweicloud_cse_microservice_instance/huaweicloud_cse_microservice

File: huaweicloud/services/cse/resource_huaweicloud_cse_microservice_instance.go [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+		Status:         d.Get("status").(string),
+		Properties:     d.Get("properties").(map[string]interface{}),
+		HealthCheck:    buildHealthCheckStructure(d.Get("health_check").([]interface{})),
+		DataCenterInfo: buildDataCenterStructure(d.Get("data_center").([]interface{})),
+	}
+}
+
+func resourceMicroserviceInstanceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var token string
+	if user, ok := d.GetOk("admin_user"); ok {
```
Comment (hmbig2) [inline huaweicloud/services/cse/resource_huaweicloud_cse_microservice_instance.go]: extract function: line 187 to line 196

File: docs/resources/cse_microservice.md (hunk 2 above)
Comment (Lance52259) [reply docs/resources/cse_microservice.md:80]: the engine resource have auth_type parameter.

File: docs/resources/cse_microservice.md (hunk 3 above)
Comment (Lance52259) [reply docs/resources/cse_microservice.md:83]: ditto.

File: huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go (hunk 4 above)
Comment (Lance52259) [reply huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go]: This function is to generate the customized client and the API version is actually v4 in documentation.

File: docs/resources/cse_microservice_instance.md (hunk 7 above)
Comment (Lance52259) [reply docs/resources/cse_microservice_instance.md]: sure, that is wrong. BTW, is `s/huaweicloud_cse_microservice/huaweicloud_cse_microservice_instance`

File: docs/resources/cse_microservice.md (hunk 1 above)
Comment (niuzhenguo) [inline docs/resources/cse_microservice.md]: Defaults to `UP`

Comment (Lance52259) [reply docs/resources/cse_microservice.md]: done.

File: huaweicloud/services/cse/resource_huaweicloud_cse_microservice_instance.go (hunk 8 above)
Comment (Lance52259) [reply huaweicloud/services/cse/resource_huaweicloud_cse_microservice_instance.go]: done.

File: docs/resources/cse_microservice.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  The password format must meet the following conditions:
+  + Must be `8` to `32` characters long.
+  + A password must containing at least one digit, one uppercase letter, one lowercase letter, and one special character
```
Comment (niuzhenguo) [inline docs/resources/cse_microservice.md]: Nits: s/containing/contain

File: docs/resources/cse_microservice.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+  + Cannot be the account name or account name spelled backwards.
+  + The password can only start with a letter.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The microservice ID.
+
+* `status` - The microservice status. The values supports **UP** and **DOWN**.
```
Comment (niuzhenguo) [inline docs/resources/cse_microservice.md:97]: `status` already included in Arguments above.

File: huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+				ForceNew: true,
+			},
+			"admin_pass": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Sensitive:    true,
+				ForceNew:     true,
+				RequiredWith: []string{"admin_user"},
+			},
+			"status": {
```
Comment (niuzhenguo) [inline huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go:88]: as described in the docs, this is an argument in stead of attribute?

File: huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+	deleteOpts := services.DeleteOpts{
+		Force: true,
+	}
+	client := NewCustomClient(d.Get("connect_address").(string), "v4", "default")
+	err = services.Delete(client, deleteOpts, d.Id(), token)
+	if err != nil {
+		return diag.Errorf("error deleting dedicated microservice (%s): %s", d.Id(), err)
+	}
+
+	d.SetId("")
```
Comment (niuzhenguo) [inline huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go:173]: Nits: this is unneeded.

File: docs/resources/cse_microservice.md (hunk 1 above)
Comment (Lance52259) [reply docs/resources/cse_microservice.md]: Actually it is an attribute, I forgot to delete.

File: docs/resources/cse_microservice.md (hunk 9 above)
Comment (Lance52259) [reply docs/resources/cse_microservice.md]: done.

File: docs/resources/cse_microservice.md (hunk 10 above)
Comment (Lance52259) [reply docs/resources/cse_microservice.md:97]: It is an attribute, the description of the arguments part has been deleted.

File: huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go (hunk 11 above)
Comment (Lance52259) [reply huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go:88]: Attribute is right.

File: huaweicloud/services/acceptance/cse/resource_huaweicloud_cse_microservice_instance_test.go [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+resource "huaweicloud_cse_microservice" "test" {
+  connect_address = huaweicloud_cse_microservice_engine.test.service_registry_addresses.0.public
+
+  name        = "%[2]s"
+  app_name    = "%[2]s"
+  environment = "development"
+  version     = "1.0.1"
+  level       = "BACK"
+  status      = "UP"
```
Comment (niuzhenguo) [inline huaweicloud/services/acceptance/cse/resource_huaweicloud_cse_microservice_instance_test.go]: status already changed to attribute

File: huaweicloud/services/cse/resource_huaweicloud_cse_microservice_instance.go [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"properties": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
```
Comment (niuzhenguo) [inline huaweicloud/services/cse/resource_huaweicloud_cse_microservice_instance.go:63]: we have an API to support updating properties.

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/cse' TESTARGS='-run=TestAccMicroservice_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cse -v -run=TestAccMicroservice_basic -timeout 360m -parallel 4
=== RUN   TestAccMicroservice_basic
=== PAUSE TestAccMicroservice_basic
=== CONT  TestAccMicroservice_basic
--- PASS: TestAccMicroservice_basic (886.95s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cse      887.118s
```

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/cse' TESTARGS='-run=TestAccMicroserviceInstance_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cse -v -run=TestAccMicroserviceInstance_basic -timeout 360m -parallel 4
=== RUN   TestAccMicroserviceInstance_basic
=== PAUSE TestAccMicroserviceInstance_basic
=== CONT  TestAccMicroserviceInstance_basic
--- PASS: TestAccMicroserviceInstance_basic (876.20s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cse       876.290s
```

File: huaweicloud/services/acceptance/cse/resource_huaweicloud_cse_microservice_instance_test.go (hunk 13 above)
Comment (Lance52259) [reply huaweicloud/services/acceptance/cse/resource_huaweicloud_cse_microservice_instance_test.go]: done.

File: docs/resources/cse_microservice_instance.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+  Changing this will create a new microservice instance.
+
+* `data_center` - (Optional, List, ForceNew) Specifies the data center configuration.
+  The [object](#microservice_instance_data_center) structure is documented below.
+  Changing this will create a new microservice instance.
+
+* `admin_user` - (Optional, String, ForceNew) Specifies the account name. The initial account name is **root**.
+  Required if the `auth_type` of engine is **RBAC**. Changing this will create a new microservice instance.
+
+-> Please make sure that the public IP has been bound to the Engine before use.
```
Comment (ShiChangkuo) [inline docs/resources/cse_microservice_instance.md]: it seems that this line has no relationship with the context?

File: docs/resources/cse_microservice_instance.md [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+* `status` - The microservice instance status. The values supports **UP**, **DOWN**, **STARTING** and **OUTOFSERVICE**.
+
+## Import
+
+Microservices can be imported using related `connect_address`, `microservice_id` and their `id`, separated by a
+slash (/), e.g.
+
+```
+$ terraform import huaweicloud_cse_microservice_instance.test https://124.70.26.32:30100/f14960ba495e03f59f85aacaaafbdef3fbff3f0d
```
Comment (ShiChangkuo) [inline docs/resources/cse_microservice_instance.md]: the \<instance id\> is missing in the example

File: huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+
+	var mErr *multierror.Error
+	formatErr := fmt.Errorf("The imported microservice ID specifies an invalid format, must be " +
+		"<cnnect_address>/<microservice_id> or <cnnect_address>/<microservice_id>/<admin_user>/<admin_pass>.")
+
+	resp := re.FindAllStringSubmatch(d.Id(), -1)
+	if len(resp) >= 1 && len(resp[0]) == 3 {
+		mErr = multierror.Append(mErr, d.Set("connect_address", resp[0][1]))
+		parts := strings.Split(resp[0][2], "/")
+		if len(parts) >= 1 {
```
Comment (ShiChangkuo) [inline huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go]: using `switch` will be more clear?
```
	switch len(parts) {
	case 1:

	case 3:

	default:
		
	}
```

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/cse' TESTARGS='-run=TestAccMicroservice_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cse -v -run=TestAccMicroservice_basic -timeout 360m -parallel 4
=== RUN   TestAccMicroservice_basic
=== PAUSE TestAccMicroservice_basic
=== CONT  TestAccMicroservice_basic
--- PASS: TestAccMicroservice_basic (923.12s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cse   923.257s
```

File: huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go (hunk 17 above)
Comment (Lance52259) [reply huaweicloud/services/cse/resource_huaweicloud_cse_microservice.go]: sure, that will be more clear.

File: docs/resources/cse_microservice_instance.md (hunk 16 above)
Comment (Lance52259) [reply docs/resources/cse_microservice_instance.md]: done.

File: docs/resources/cse_microservice_instance.md (hunk 15 above)
Comment (Lance52259) [reply docs/resources/cse_microservice_instance.md]: done.

File: huaweicloud/services/cse/resource_huaweicloud_cse_microservice_instance.go (hunk 14 above)
Comment (Lance52259) [reply huaweicloud/services/cse/resource_huaweicloud_cse_microservice_instance.go:63]: Support in the next PR.

**Comment by Lance52259** [comment]
```
make testacc 'EST='./huaweicloud/services/acceptance/cse' TESTARGS='-run=TestAccMicroserviceInstance_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cse -v -run=TestAccMicroserviceInstance_basic -timeout 360m -parallel 4
=== RUN   TestAccMicroserviceInstance_basic
=== PAUSE TestAccMicroserviceInstance_basic
=== CONT  TestAccMicroserviceInstance_basic
--- PASS: TestAccMicroserviceInstance_basic (896.64s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cse   896.741s
```

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/cse' TESTARGS='-run=TestAccMicroservice_basic'==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cse -v -run=TestAccMicroservice_basic -timeout 360m -parallel 4
=== RUN   TestAccMicroservice_basic
=== PAUSE TestAccMicroservice_basic
=== CONT  TestAccMicroservice_basic
--- PASS: TestAccMicroservice_basic (889.57s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cse    889.687s
```
```
make testacc TEST='./huaweicloud/services/acceptance/cse' TESTARGS='-run=TestAccMicroserviceInstance_basic'rvices/acceptance/cse' TESTARGS='-run=TestAccMicroservi==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cse -v -run=TestAccMicroserviceInstance_basic -timeout 360m -parallel 4
=== RUN   TestAccMicroserviceInstance_basic
=== PAUSE TestAccMicroserviceInstance_basic
=== CONT  TestAccMicroserviceInstance_basic
--- PASS: TestAccMicroserviceInstance_basic (894.08s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cse      894.233s
```

**Pull request merged**